src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m

```diff
@@ -143,10 +229,537 @@ + (JavaComponentAccessibility *) getComponentAccessibility:(NSString *)role
     if (className != nil) {
         return [NSClassFromString(className) alloc];
     }
-    return nil;
+    return [CommonComponentAccessibility alloc];
+}
+
++ (CommonComponentAccessibility *) getComponentAccessibility:(NSString *)role andParent:(CommonComponentAccessibility *)parent
+{
+    AWT_ASSERT_APPKIT_THREAD;
+    if (rolesMap == nil) {
+        [self initializeRolesMap];
+    }
+    NSString *className = [rowRolesMapForParent objectForKey:[[parent class] className]];
+    if (className == nil) {
+        return [CommonComponentAccessibility getComponentAccessibility:role];
+    }
+    return [NSClassFromString(className) alloc];
+}
+
+- (id)initWithParent:(NSObject *)parent withEnv:(JNIEnv *)env withAccessible:(jobject)accessible withIndex:(jint)index withView:(NSView *)view withJavaRole:(NSString *)javaRole
+{
+    self = [super init];
+    if (self)
+    {
+        fParent = [parent retain];
+        fView = [view retain];
+        fJavaRole = [javaRole retain];
+
+        fAccessible = (*env)->NewWeakGlobalRef(env, accessible);
+        (*env)->ExceptionClear(env); // in case of OOME
+        jobject jcomponent = [(AWTView *)fView awtComponent:env];
+        fComponent = (*env)->NewWeakGlobalRef(env, jcomponent);
```

Comment:
What happens if "fAccessible" and "fComponent" will be null due to OOM, should we rethrow this exception, or ignoring it is fine?

Reply:
I added check for exceptions.